File: SudokuSolver.java
import java.util.Scanner;

public class SudokuSolver {
    public static void solveSudoku(char[][] board) {
        helper(board, 0, 0);
    }

    // function to check is the number is safe in particular cell
    public static boolean isSafe(char[][] board, int row, int col, int number) {
        // rows and columns
        for (int i = 0; i < board.length; i++) {
            if (board[row][i] == (char) (number + '0')) {
                return false;
            }
            if (board[i][col] == (char) (number + '0')) {
                return false;
            }
        }

        int sr = (row / 3) * 3;
        int sc = (col / 3) * 3;
        for (int i = sr; i < sr + 3; i++) {
            for (int j = sc; j < sc + 3; j++) {
                if (board[i][j] == (char) (number + '0')) {
                    return false;
                }
            }
        }
        return true;

    }

    public static boolean helper(char[][] board, int row, int col) {
        if (row == board.length) {
            return true;
        }
        int nrow = 0;
        int ncol = 0;
        if (col != board.length - 1) {
            nrow = row;
            ncol = col + 1;
        } else {
            nrow = row + 1;
            ncol = 0;
        }

        if (board[row][col] != '.') {
            if (helper(board, nrow, ncol)) {
                return true;
            }
        } else {
            for (int i = 1; i <= 9; i++) {
                if (isSafe(board, row, col, i)) {
                    board[row][col] = (char) (i + '0');
                    if (helper(board, nrow, ncol)) {
                        return true;
                    } else {
                        board[row][col] = '.';
                    }
                }
            }
        }

        return false;
    }

    static void printSudoku(char[][] board) {
        System.out.println("-------------------------------------");
        for (int i = 0; i < 9; i++) {
            System.out.print("| ");
            for (int j = 0; j < 9; j++) {
                System.out.print(board[i][j] + " | ");
            }
            System.out.println();
            System.out.println("-------------------------------------");
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        char[][] board = new char[9][9];
        System.out.println("Enter Input: ");
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                board[i][j] = sc.next().charAt(0);
            }
        }
        printSudoku(board);
        solveSudoku(board);
        System.out.println("Output");
        printSudoku(board);
        sc.close();
    }
}
/*
 * Input:
 * 5 3 . . 7 . . . .
 * 6 . . 1 9 5 . . .
 * . 9 8 . . . . 6 .
 * 8 . . . 6 . . . 3
 * 4 . . 8 . 3 . . 1
 * 7 . . . 2 . . . 6
 * . 6 . . . . 2 8 .
 * . . . 4 1 9 . . 5
 * . . . . 8 . . 7 9
 */

File: SudokuSolver.cpp
#include <iostream>
#include <vector>
#define rep(i, a, b) for (int i = a; i < b; i++)
#define nl cout << endl
using namespace std;

bool isSafe(int grid[9][9], int row, int col, int num)
{
    rep(i, 0, 9)
    {
        if (grid[i][col] == num)
        {
            return false;
        }
        if (grid[row][i] == num)
        {
            return false;
        }
    }
    int sr = (row / 3) * 3;
    int sc = (col / 3) * 3;
    rep(i, sr, sr + 3)
    {
        rep(j, sc, sc + 3)
        {
            if (grid[i][j] == num)
            {
                return false;
            }
        }
    }
    return true;
}

bool helper(int grid[9][9], int row, int col)
{
    if (row == 9)
    {
        return true;
    }

    int nrow = 0, ncol = 0;
    if (col != 8)
    {
        ncol = col + 1;
        nrow = row;
    }
    else
    {
        nrow = row + 1;
        ncol = 0;
    }

    if (grid[row][col] != 0)
    {
        if (helper(grid, nrow, ncol))
        {
            return true;
        }
    }
    else
    {
        rep(i, 1, 10)
        {
            if (isSafe(grid, row, col, i))
            {
                grid[row][col] = i;
                if (helper(grid, nrow, ncol))
                {
                    return true;
                }
                else
                {
                    grid[row][col] = 0;
                }
            }
        }
    }
    return false;
}

void solveSudoku(int grid[9][9])
{
    helper(grid, 0, 0);
}
void printSudoku(int grid[9][9])
{
    cout << "------------------------------------" << endl;
    for (int i = 0; i < 9; i++)
    {
        cout << "|";
        for (int j = 0; j < 9; j++)
        {
            cout << grid[i][j] << " | ";
        }
        nl;
        cout << "------------------------------------" << endl;
    }
}
int main()
{
    int grid[9][9];
    cout << "Enter Input" << endl;
    for (int i = 0; i < 9; i++)
    {
        for (int j = 0; j < 9; j++)
        {
            cin >> grid[i][j];
        }
    }

    cout << "Entered Sudoku: " << endl;
    printSudoku(grid);
    solveSudoku(grid);
    cout << "Output: " << endl;
    printSudoku(grid);
    return 0;
}

/*
Input
3 0 6 5 0 8 4 0 0
5 2 0 0 0 0 0 0 0
0 8 7 0 0 0 0 3 1
0 0 3 0 1 0 0 8 0
9 0 0 8 6 3 0 0 5
0 5 0 0 9 0 6 0 0
1 3 0 0 0 0 2 5 0
0 0 0 0 0 0 0 7 4
0 0 5 2 0 6 3 0 0
*/
